USWDS - Input mask: Version 2

amyleadem opened the pull request:
# Summary
**Updated input mask functionality and attributes.** This update simplifies the mask attribute requirements, provides instant feedback to the user if they enter an invalid character, and allows for better keyboard controls.

###  Some notes:
- The original branch for this work (`mm-input-mask`) in PR https://github.com/uswds/uswds/pull/5104 appears to have been corrupted and/or is missing commits. This PR attempts to fill in the gaps in history and resolve items that were identified in the comments of that PR. 
- I am proposing that this PR acts as an MVP and includes (at least) all possible breaking changes. There are some items that need more JavaScript work and I have outlined them in  "Outstanding issues"  below. I don't believe any of these would be breaking changes and I propose we open new issues for these items to keep things moving.

## Breaking change
:warning: This is a breaking change. 
1. The input must now include the `usa-input-mask` class. Before the class was named `usa-masked`.
1. Required attributes have been updated. See "Component attribute updates" table below for details. 
2. The `mask` attribute requires different values than `pattern` and `data-charset` did before. See "Component attribute updates" table below for details. 
 
## Related issues

Closes #5054, closes #5227. 

Documentation update issue: https://github.com/uswds/uswds-site/issues/2062

## Related pull requests
This work originates from PR https://github.com/uswds/uswds/pull/5104

Documentation update (WIP): https://github.com/uswds/uswds-site/pull/2073

## Preview link

Preview link: [Input mask](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-input-mask/?path=/story/components-form-inputs-input-mask--ssn)

## Major changes
This is a significant rewrite that:

### Improved end-user interaction
- Added a screen reader-friendly hint message to notify the user when they enter an invalid character type
- Automatically advance cursor past mask punctuation
- Prevent flashing disallowed character input
- Allow common arrow cursor interactions (up: move to end, down: move to front, etc)
- Allow leading mask punctuation (like `(` in phone numbers)
- Better handle pasted content
- Better handle multi-key combination input (shift-, option-, etc key combinations)

### Enhanced input mask creation attributes
- Simplified the method for adding a mask pattern
    - Replaced the regex-based `pattern` and `data-charset` attributes with the `mask` attribute, which accepts the letter `A` to represent any alphabetic character, the number `9` to represent any numeric character, and format characters like `-` and `/` 
        - Note: Full list of format characters can be found in [lines 16-32 of `index.js`](https://github.com/uswds/uswds/blob/c8344a5113d2bc88dffe364b202efe5844655a44/packages/usa-input-mask/src/index.js#L16-L32).
- New invalid character hint message can be customized with data attributes
    - Users can now customize status message text with new `data-invalid-alpha-text` and `data-invalid-numeric-text` attributes
- Added ability to auto-format casing with forced-case attributes
    - Added new `forceupper` and `forcelower` attributes
    
### Improved developer experience
- Added unit tests
- Added JSDocs comments to each function

### Component attribute updates
|  Attribute name  | Previous name | Required? | Description |
| :--- | :--- | :--- | :--- |
| .usa-input-mask | .usa-masked | Yes | Triggers component initialization and adds styles |
| data&#8209;invalid&#8209;alpha&#8209;text | -- | No | Overrides the default value for the invalid alphabetic character hint text. Default value: `Please enter a letter character here` |
| data&#8209;invalid&#8209;numeric&#8209;text | -- | No | Overrides the default value for the invalid number character hint text. Default value: `Please enter a number character here` |
| forcelower | -- | No | Forces casing on alphabetic characters. Add `forcelower="true"` to the input element to enforce lowercase letters. |
| forceupper | -- | No | Forces casing on alphabetic characters. Add `forceupper="true"` to the input element to enforce uppercase letters. |
| inputmode | inputmode | No | Tells browsers which keyboard type to display. For masks that contain all numbers, set `inputmode` to numeric to display a numeric keyboard. Set `inputmode` to `tel` to display a telephone keyboard. Set `inputmode` to `text` if your input mask includes alphabetic characters. |
| mask   | data&#8209;charset, pattern | Yes | Defines the format of your mask. Use the number `9` to represent any number, the letter `A` to represent any letter, and any format character needed in your mask. For example, an input that requires three numbers followed by four letters with a hyphen in-between would have a  `mask` attribute that looks like `999-AAAA`. |
| placeholder  | placeholder | Yes |  Creates the visual placeholder and informs some elements of mask formatting. Use underscores `_` for character input placeholders combined with any valid format characters to customize your input mask. For example, a U.S. telephone `placeholder` attribute may look like `(___) ___-____`. |

## Problem statement
1. The input mask component does not provide any feedback when an invalid character is entered.
3. Keyboard interactions result in unpredictable outcomes
4. Setting up the input mask component pattern can seem complicated
5. Using format characters can cause unpredictable outcomes

Comment from PR [#5104](https://github.com/uswds/uswds/pull/5104#issuecomment-1403635185):
> Here are some of the details on why a rewrite was initiated:
> - error/invalid character handling
>   - the new mask gives instant feedback for invalid characters, similar to how the character counter behaves
>   - btw I saw the issues you found with spaces which I will address first
> - weirdness with the current mask https://designsystem.digital.gov/components/input-mask/
>   - when you type an invalid character, it shows the character for a split second and then disappears (screen readers call it out as well, this was perhaps the biggest issue)
>   - with the current mask, there is some weirdness in a field with format characters (like telephone)... move to the right of a format character and try to use the arrow keys
>   - you can't highlight all and delete on some masks (like telephone)
>   - you can't use a leading format character
>     - so with the current mask, you can't have (___) ___-____ we were forced to go with ___-___-____

## Solution

### Tasks completed since #5104 
- [x] Add `forceupper` and `forcelower` attribute controls to storybook ([53a0b8a](https://github.com/uswds/uswds/pull/5227/commits/53a0b8adc965146c10a4b278c397eee7a0dd877c))
- [x] Remove default status message ([5f55796](https://github.com/uswds/uswds/pull/5227/commits/5f55796f4a76dd1ebd4690dd16232408f189554b))
- [x] Remove `usa-input-mask__field` markup requirement
    - Moved `usa-input-mask` from container element to the input itself. This replaced the need for `.usa-input-mask__field`. (https://github.com/uswds/uswds/pull/5227/commits/9a49a4e5aecb78d08a99d09aa5a7b4e4c42ab898)
- [x] Dynamically associate `aria-describedby`with SR status message element
- [x] Removed  the need for empty status markup at init
    - Dynamically generates the status message element instead ([5f55796](https://github.com/uswds/uswds/pull/5227/commits/5f55796f4a76dd1ebd4690dd16232408f189554b))
- [x] Align use of `usa-form` and `usa-form-group` with other form components
    - Originated from this comment: https://github.com/uswds/uswds/pull/5104#discussion_r1084628176
    - Moved `usa-input-mask` from a container element to the input element itself. This means that `usa-form` and `usa-form-group` are no longer contained inside `usa-input-mask`. (https://github.com/uswds/uswds/pull/5227/commits/9a49a4e5aecb78d08a99d09aa5a7b4e4c42ab898)
- [x] Align error message styling with other components
    - Replaced custom error text styles with `usa-error-message` class ([a16a662](https://github.com/uswds/uswds/pull/5227/commits/a16a662514af3e8198521040dc4efd2cb60b234e))
- [x] Add data attributes to customize status message languages ([2b4e778](https://github.com/uswds/uswds/pull/5227/commits/2b4e778e5a632384fba23ae67f8a056cd09effc1))
- [x] Assess status message language (Should we find a less passive alternative?)
    - Changed "A number character is required here" to "Please enter a number here" (Thanks @bonnieAcameron @sarah-sch for the assist)
- [x] Remove date story
    - Removed date story because input mask cannot successfully validate dates (Example: it is possible to submit 99/99/9999 without error) ([1e70efa](https://github.com/uswds/uswds/pull/5227/commits/1e70efac844ace54fdb2bfdc30a45c89b6e4c2a9))
- [x] Fix - cannot tab to next item ([aa27c1f](https://github.com/uswds/uswds/pull/5227/commits/aa27c1fd5b36b95db8833ffb6cf99b68e650c146))
    - We should confirm that this fix doesn't break anything.
- [x] Checks for required attributes:`mask`, `placeholder` ([ede9b6f](https://github.com/uswds/uswds/pull/5227/commits/ede9b6fbbddd744184322e33a4271ca3d547c74b))
- [ ] Trigger invalid input hint with each invalid character, instead of just the first
    - We should add a default status so that screen reader reads out more consistently
    - Depending on the content of the default status, we might want to consider removing the `aria-describedby` connection done in https://github.com/uswds/uswds/pull/5227/commits/8f4f64619f64a64cfda9109f1ebd1853169084da
- [ ] In numeric masks, special characters are not identified as invalid
    - The following characters do not trigger the invalid character warning AND appear in the **_numeric_** mask: ! # ^ &
    - The following characters do not trigger the invalid character warning and do NOT appear in the _**numeric**_ mask: @ $ % * ( ) -
        - Note: these are included in `FORMAT_CHARACTERS`
- [ ] In alphanumeric masks, special characters can trigger the incorrect invalid warning
    - For example, in the [alphanumeric example](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-input-mask/iframe.html?id=components-form-inputs-input-mask--alphanumeric&viewMode=story), typing `@` into the first position triggers a "Please enter a valid number character here", but the pattern is expecting a letter. 
- [ ] Pasting in content with invalid characters does nothing and provides no feedback. Example: pasting "145-62-0495" into the alphanumeric example
- [ ] Pasting in partial content with valid characters does nothing and provides no feedback. For example in the [alphanumeric example](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-input-mask/iframe.html?id=components-form-inputs-input-mask--alphanumeric&viewMode=story), type the letter `A` into the first position, then try copying the string `BC1234` or `BC-1234` to finish it out. 
- [ ] Multi-selection then change/delete does not work. Multi-selection then delete deletes the character before the selection not the selected text. Note that select-all then delete does work as expected.
- [ ] `Cut`/`ctrl` + `x` or `ctrl` + `backspace` removes chunk of placeholder 
- [ ] ~Add clarity to hint examples~
    - This will be handled in issue https://github.com/uswds/uswds/issues/5244
- [ ] Update unit tests
- [ ] Make code more readable/Add JS comments

### Open discussion items
From @thisisdano in https://github.com/uswds/uswds/pull/5227#issuecomment-1507445226):
- Should there be a "Please enter a valid character" message pre-populating the status field? 
    - Dan: I'm inclined to say that this should be blank.
    - Amy: I removed this, but realized that the default message was preserving the vertical space. Now the content below will get bumped down if an invalid character is entered. 
- Disallowed character status updates announce only once in VO. Should they run after each input of a disallowed character? 
    - Dan: I think yes
    - Amy: At least in voiceover, it would help to have a hint message that serves as the screen reader default status so that the message element isn't empty after invalid selection. 
- Should there be a status update for trying to enter more characters than the mask supports (i.e. every time you try to enter a character beyond the mask limit)? 
    - Dan: I think yes
    - Amy: I have added this to "Outstanding items"
- And a big one: Should input mask be a separate "component" or a data-element driven enhancement on the standard text input? Or something else?
    - Amy: TBD - we can discuss in the comments and at dev sync if not resolved before that

Other questions :
- I removed the date example story because it wasn't able to mask all invalid characters (example: 99/99/9999 is a valid date). Did we want to keep the date story as is, improve it so it has limitations on which numbers are expected in each field, or just rely on the date picker component to do this? 
- Are there any other breaking changes we want to include? Is this the ideal structure?

### Outstanding items 
_Proposing we open new issues for these_
- Entering a double-space creates a `.` in MacOS and causes a break in the input
    - Problem identified in https://github.com/uswds/uswds/pull/5104#discussion_r1085830891
    - Possible solution identified in https://github.com/uswds/uswds/pull/5104#discussion_r1091387732
- Users are not notified when `maxlength` has been reached
- Phone number example reads `(` in voiceover

## Testing and review
Confirm that the component meets functionality requirements as defined in the [Input Mask requirements document](https://docs.google.com/document/d/1kLHaCl5PceSByz0NvZqSc-ot7tSjcbe1EJPbBAZM7LY/edit#) (Google docs) :lock:

mitchmoccia committed:
Commit package-lock

mitchmoccia committed:
Merge branch 'jb-inclusive-patterns-input-mask' of https://github.com/uswds/uswds into jb-inclusive-patterns-input-mask

mitchmoccia committed:
Commit package-lock

mitchmoccia committed:
Rewrite input mask to address several requested features

mitchmoccia committed:
Input mask rewrite in progress updates

mitchmoccia committed:
Implement dynamic invalid masked character messaging and more refactoring

mitchmoccia committed:
Implement shell mask feature so when typing, mask characters are replaced with typed characters

mitchmoccia committed:
Input mask rewrite in progress updates

mitchmoccia committed:
Input mask rewrite in progress updates

mitchmoccia committed:
Input mask rewrite in progress updates

mitchmoccia committed:
Address pasted data and arrow key navigation issues

mitchmoccia committed:
Fix mask issue where paste would allow letters in number mask and other refactorings

mitchmoccia committed:
Refactor and clean up

mitchmoccia committed:
Update Input mask unit test

mitchmoccia committed:
Finalize the rest of the unit tests for input mask

mitchmoccia committed:
Add keyup handler, prevent arrow key to left of leading format character  and more refactoring

mitchmoccia committed:
Finalize unit tests for input mask

mitchmoccia committed:
Allow select all quick key in field plus fix issue with field history allowing letters in a numeric only mask

mitchmoccia committed:
Fix incorrect id on date short and clean up

mitchmoccia committed:
Fix style related issue with leading format characters, plus update unit tests

mitchmoccia committed:
Merge in develop and resolve conflicts

mitchmoccia committed:
Update package-lock

mitchmoccia committed:
Fix CSS for input mask shell formatting

mitchmoccia committed:
Implement input mask unit tests that check for correct mask formatting

mitchmoccia committed:
Clean up input mask unit tests and fix a hardcoded querySelector in the component

mitchmoccia committed:
Fix issue with input mask date unit test

mitchmoccia committed:
Remove the pattern attribute from all input mask examples

mitchmoccia committed:
Remove the pattern attribute from input unit tests

mitchmoccia committed:
Change US to U.S. for input mask examples plus one update for this in the table component

mitchmoccia committed:
Set forceupper to true without the quotes

mitchmoccia committed:
Change hint text for date input mask from 01-01-1970 to 01-31-1970

Commit:
Merge develop

Commit:
Revert package-lock

Commit:
Fix typo - Change 'defaulVal' to 'defaultVal'

thisisdano commented:
I'm going to make a running list of the features in this PR, and I'll keep adding to this comment as I identify them.

- Provide feedback for disallowed characters
- Automatically advance cursor past mask punctuation
- Prevent flashing disallowed character input
- Allow common arrow cursor interactions (up: move to end, down: move to front, etc)
- Allow leading mask punctuation (like `(` in phone numbers)
- Better handle pasted content
- Better handle multi-key combination input (shift-, option-, etc key combinations)
- Auto-format casing for forced-case letters

thisisdano commented:
I'm going to make a running list of what I see as some of the issues in this work:

- Multi-selection then change/delete does not work. Multi-selection then delete deletes the character _before_ the selection not the selected text.
- I'm not sure if there should be a `Please enter a valid character` message pre-populating the status field. I'm inclined to say that this should be blank.
- Disallowed character status updates announce only once in VO. Should they run after each input of a disallowed character?
- Should there be a status update for trying to enter more characters than the mask supports? ie every time you try to enter a character beyond the mask limit.

Commit:
Add forcelower/forceupper controls in storybook

Commit:
Remove default message; remove usa-input-mask__message from twig

Commit:
Run prettier

Commit:
Remove usa-input-mask__message from all templates; remove usa-input-mask__message unit tests

Commit:
Replace usa-input-mask__field with usa-input-mask

Commit:
Replace usa-input-mask__field with usa-input-mask

Commit:
Remove requirement to add a container in the markup

- Rename variables - share status message elements across functions -m Modify getMaskedElements

Commit:
Replace custom classes with usa-error-message and usa-sr-only

Commit:
Use data attributes to customize status message text

Commit:
Fix forceupper and forcelower attributes in twig

Commit:
Fix broken tab navigation

Commit:
Fix broken unit tests

Commit:
Merge branch 'develop' of https://github.com/uswds/uswds into al-input-mask

Commit:
Remove date short story

Commit:
Revert SSN hint text to original

Commit:
Update invalid character text

Make text less passive

Commit:
Update invalid number text in unit tests

Commit:
Remove hint-info markup from templates

Commit:
Update phone inputmode to 'tel'

Commit:
Add test story for custom hint text and multiple inputs

Commit:
Remove maxlength attribute from templates

Commit:
Remove pattern attribute from main template

Commit:
Run prettier; format twig

amyleadem commented:
Pulling in this[ detailed comment](https://github.com/uswds/uswds/pull/5104#pullrequestreview-1266406466) from @mejiaj in PR #5104 to make sure we address all items:

Submitting initial review for functionality. Overall I like the new features and I'm curious as to the reasons for the rewrite. The JS especially has gone through a lot of changes.

As a developer what changes would I need to make (if any) to upgrade to this version? Do we have that available as guidance?

Some standout items from notes:
1. I like that you can't paste invalid characters into SSN.
2. Entering <kbd>Space</kbd> multiple times has weird results (example in comments).
3. Status messages disappear when switching between variants (maybe storybook bug).
4. Default status message doesn't return once input has returned to "blank" state (Met input requirements, but then deleted them to go back to initial state)
5. Microsoft edge on windows shows HTML form message `Please match the requested format` when filled out (probably something with the pattern?)
6. Trying to delete chunks of text via <kbd>ctrl+backspace</kbd> on windows makes the input mask disappear. You can reproduce in OSX too by holding the same key combination.
7. **NVDA** quirks on reading "U.S."
8. **NVDA** reads SSN numbers as hundreds/thousands (when reading hint).
9. **NVDA** alphanumeric doesn't read anything on page load or tab/focus.

I tested general functionality here on OSX, Windows, and NVDA (windows, microsoft edge).

Commit:
Style update: align __content color with usa-hint; remove unnecessary rules for usa-input-mask

Commit:
Add empty space to srStatusMessage when hiding message

This encourages VoiceOver to read out the text change

Commit:
Add SR status message element to input's aria-describedby

amyleadem commented:
@mejiaj, @mahoneycm 
Tagging you for initial review of this refactor. It could use some sharp eyes on the JavaScript work. 
The PR description includes a lot of detail in an attempt to explain everything that has been done here, but please feel free to reach out if you have any questions.

mahoneycm commented on packages/usa-input-mask/src/content/usa-input-mask-custom-text.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "customInvalidAlphaText": "Ingrese un carácter de letra aquí",
+  "customInvalidNumericText": "Ingrese un carácter de número aquí",
```
Setting custom invalid text works great and is super easy! I experimented with other templates and everything went smoothly

amyleadem replied:
Note that the Spanish might need tweaking. I was just trying to apply a realistic use case.

mahoneycm commented on packages/usa-input-mask/src/content/usa-input-mask-phone.json:
```diff
@@ -1,9 +1,9 @@
+  "label": "U.S. Telephone Number",
+  "mask": "(999) 999-9999",
+  "placeholder": "(___) ___-____",
+  "type": "tel"
```
I tried experimenting with changing the input `type` and opening the keyboard on mobile but nothing changed.

I'm not sure if there is another assistive technology that this is utilized with but wanted to flag it!

mahoneycm commented on packages/usa-input-mask/src/index.js:
```diff
@@ -1,124 +1,1095 @@
+  const isNumber = /^\d+$/.test(key);
+  const invalidAlpha = inputEl.getAttribute("data-invalid-alpha-text")
+    ? inputEl.getAttribute("data-invalid-alpha-text")
+    : "Please enter a letter character here";
+  const invalidNumeric = inputEl.getAttribute("data-invalid-numeric-text")
+    ? inputEl.getAttribute("data-invalid-numeric-text")
+    : "Please enter a number character here";
```
- Should the error message specify the user entered an invalid input? Maybe something like
 
> Invalid character. Please enter a _alpha/number_ character here

amyleadem replied:
I like this suggestion. I'll incorporate tomorrow unless I see any objections.

mahoneycm reviewed (commented):
So far, I'm loving this update Amy! I still need to do a deep dive into the JS but wanted to bring up a couple of things I noticed in my initial testing.

## Accessibility testing

With VoiceOver, I noticed the invalid comment does not get read after each invalid input as expected.
- When multiple invalid inputs are entered in succession, it only reads the invalid character message after the first
- If you enter an invalid character, followed by a valid character, then _another_ invalid character, VoiceOver does not read the new error

### Other questions

Some thoughts I had I wanted to throw into the ether! 

- Special characters such as emojis or symbols do not trigger error
- I wondered if the input that received the error should receive the error border to add clarity. Input labels add top margin are easy to distinguish from one another, but I'm wondering how side-by-side inputs will render. I'll try to test further!

mejiaj reviewed (changes requested):
Nice job so far. I've just got a couple comments/questions on a few outstanding items.

## PR description

1. Components attributes table 
2. Open discussion & outstanding items

### 1. Components attributes table 

The table Component attribute updates is helpful, but can be hard to read because of alignment and classes wrapping.

Here's what I see:
![image](https://user-images.githubusercontent.com/3385219/234379451-1541c0fc-4f90-4bf1-b08f-8dfa70933bee.png)


### 2. Open discussion & outstanding items

I see the PR description was last edited two weeks ago. Are these still valid?

---

## Functionality

1. Alphanumeric example is hard to follow, can we get a more realistic example?
2. Issues with cut/paste. Mainly, when you cut text and leave the mask disappears. We should look into keyup/focus behavior to see why mask isn't being reset.

[Kapture 2023-04-25 at 13.57.09.webm](https://user-images.githubusercontent.com/3385219/234589914-c9d3162e-a77b-46dd-97d4-1d9af25ee204.webm)

## How I test

- [x] AXE a11y scans
- [x] Tested markup (missing tags/attributes/values)
- [x] Checked initial functionality
- [x] Reviewed and tested code starting at init and then behaviors

mejiaj commented on packages/usa-input-mask/src/index.js:
```diff
@@ -1,124 +1,1095 @@
+  const srStatusMessageEl = document.createElement("div");
+
+  statusMessageEl.classList.add(`${STATUS_MESSAGE_CLASS}`);
+  srStatusMessageEl.classList.add(
+    `${STATUS_MESSAGE_SR_ONLY_CLASS}`,
+    "usa-sr-only"
+  );
```
We should also have a default message. Maybe the expected state or another hint.

amyleadem replied:
> Both classes seem to be identical since `${STATUS_MESSAGE_SR_ONLY_CLASS}` is defined as `usa-sr-only` on line 14.

Removed redundant class in [9d8007b](https://github.com/uswds/uswds/pull/5227/commits/9d8007b48dc92b18a764dc4779ce87ee1ef4d685)

amyleadem commented:
@mejiaj @mahoneycm Thanks for the quick review and comments! I'll look over them today. 
FYI - The PR description was updated just a couple days ago, so all outstanding questions and notes are still valid. 
I'll add a new question about if it is safe to remove the date story and see if I can format the attributes table so it is easier to read.

Commit:
Move multiple inputs test files to test/test-patterns

Commit:
Add simpler, more realistic alphanumeric example

amyleadem commented on packages/usa-input-mask/src/index.js:
```diff
@@ -1,124 +1,1095 @@
-    input.setAttribute("data-placeholder", placeholder);
-    input.removeAttribute(`${PLACEHOLDER}`);
+
+const MESSAGE_INVALID_CLASS = `${PREFIX}-error-message`;
```
I see that there are four potential options for error styling here. Error styling can be added to `usa-form-group`, `usa-label`, `usa-input` and/or `usa-input-mask__status`. I went with just styling the status message because it is something that flagged with each invalid keystroke, instead of after the user either completes the input or form. For that reason, I wanted to make the styles as minimally disruptive as possible. Adding error styles to only the status message allows movement on only the vertical axis and not the horizontal. 

Here is what the component would look like with few different levels of error styling. Let me know if you have a preference.

![image](https://user-images.githubusercontent.com/93996430/235244239-7f1ede69-02c2-4062-ba93-cfc592337811.png)

Commit:
Fix customTextContent import in stories.js

Commit:
Remove for loop in inputMask behavior

Commit:
Throw error if input is missing mask or placeholder attributes

Commit:
Remove period from invalid message

amyleadem commented on packages/usa-input-mask/src/index.js:
```diff
@@ -1,124 +1,1095 @@
-  maskEl.textContent = "";
-  maskEl.replaceChildren(maskVal[0], maskVal[1]);
+  if (!attrs.mask) {
+    throw new Error(`${MASKED_INPUT_CLASS} is missing the mask attribute. Learn more at http://designsystem.digital.gov/components/input-mask/#available-attributes.`);
```
This links to a table that will be created when https://github.com/uswds/uswds-site/pull/2073 is merged.

amyleadem commented on packages/usa-input-mask/src/index.js:
```diff
@@ -1,124 +1,1095 @@
+
+      for (let i = 0; i < nodes.length; i += 1) {
+        options[nodes[i]] = values[i];
+      }
```
Thanks for this recommendation. This is cleaned up in https://github.com/uswds/uswds/pull/5227/commits/3b0b477b5180aef1e52730aff63b12d7b74e59d0. Let me know if you see any opportunities for improvement!

amyleadem commented on packages/usa-input-mask/src/index.js:
```diff
@@ -1,124 +1,1095 @@
+      const values = [];
+      const options = {};
+      for (
+        let attr, i = 0, attrs = maskedInput.attributes, n = attrs.length;
```
Moved the attributes definition into `enhanceInputMask()` in https://github.com/uswds/uswds/commit/3b0b477b5180aef1e52730aff63b12d7b74e59d0 when I removed the for loop in inputMask behavior. Let me know if you see any opportunities for improvement!

mejiaj replied:
Nice job!

Commit:
Remove redundant usa-sr-only class

Commit:
Create extended 'invalid character' message for SR status messages

Commit:
Change aria-live from assertive to polite

Commit:
Run lint and prettier:js

amyleadem commented on packages/usa-input-mask/src/index.js:
```diff
@@ -1,124 +1,1095 @@
+  );
+
+  statusMessageEl.setAttribute("aria-hidden", true);
+  srStatusMessageEl.setAttribute("aria-live", "assertive");
```
Updated `aria-live` to `polite` in https://github.com/uswds/uswds/pull/5227/commits/ff2b9e365e392a00838d0cfec348259fae682b01

amyleadem commented on packages/usa-input-mask/src/index.js:
```diff
@@ -1,124 +1,1095 @@
+  srStatusMessageEl.setAttribute("aria-live", "assertive");
+
+  maskedEl.insertAdjacentElement("afterend", statusMessageEl);
+  statusMessageEl.insertAdjacentElement("afterend", srStatusMessageEl);
```
Changed `aria-live` to `polite`. Next, I am going to work on some options for default messaging (at least for screen reader users)

Commit:
Remove need for statusEl params in updateMaskMessage() and hideMessage(); standardize statusEl names across functions

Commit:
Make MASK_INFO, setMaskInfo(), getMaskInfo() more readable

Rename setMaskInfo -->setPropertyValue

Rename getMaskInfo --> getPropertyValue

Make content inside functions plain language

Commit:
Rename getMatchValue --> removeFormatCharacters

Commit:
Move handlePaste to other event functions

Commit:
Update formatting & spacing

Commit:
Use more accurate name for previous getPastedText()

Commit:
Fix comment formatting

Adjustments for readability.

Commit:
Update randomly generated ID in createStatusMessages

- Allow randomID to be passed (useful in unit tests)
- Separate ID generation from string for readability

Commit:
Remove getMaskedElements()

This function returned the input element, which made it redundant.
It _used_ to return the component container, but that was removed in 9a49a4e.
.

Commit:
Validate on paste

Commit:
Add SSN example to test

Commit:
Add clarifying comments & rename minLength

Commit:
Refactor start of getValidPastedText()

- Remove redundant vars in favor of default param pastedText
- Remove loop in favor of removeFormatCharacters()
- Pass cursor position to getValidPastedText()
- Create new const to track starting point from cursor position

Commit:
Cleanup getValidPastedText(), use previous behavior for now.

amyleadem committed:
Merge pull request #5271 from uswds/jm-input-mask-paste

Input mask: validate on paste cleanup
